Design/name-mangling.md

```diff
@@ -78,111 +89,149 @@ We can identify the possible local names in this example:
-
-Let us return to the example above, with the preferred names
-annotated:
-```kotlin
-// A: A, class_A
-// x: x, field_x
-class A(val x: Int) {
-    // foo: foo, fun_foo, fun_foo_takes_Unit, fun_foo_takes_Unit_returns_Int
-    fun foo(): Int = ...
-}
+## Algorithm
 
-// B: class_B, B
-class B {
-    // get x: x, prop_x, prop_x_getter
-    // set x: x, prop_x, prop_x_setter
-    var x: Int
-        get() { ... }
-        set(v) { ... }
-}
+We work with two graphs:
 
-// A: A, fun_A, fun_A_takes_Unit, fun_A_takes_Unit_returns_Unit
-fun A() { ... }
+a **dependency graph** whose vertices are pairs 
+$(x, \text{version})$,  where $x$ is a mangled name 
+and $\text{version} \in \{\text{short}, \text{medium}, 
+\text{long}, \ldots\}$.  
+Edges encode that rendering $(x, v_x)$ requires the representation 
+of $(y, v_y)$.
 
-// foo: foo, fun_foo, fun_foo_takes_ABC, fun_foo_takes_ABC_returns_XYZ
-// x: x
-// B: B
-fun foo(x: Int, B: Int) { ... }
+a **conflict graph** with the **same set of vertices** 
+$(x, \text{version})$.  
+Edges encode that two different vertices cannot be assigned 
+simultaneously, because their resulting strings would collide.
 
-// foo: foo, fun_foo, fun_foo_takes_Unit, fun_foo_takes_Unit_returns_Unit
-fun foo() {
-    // x: x
-    val x: Int = ...
-}
-```
+### Procedure
 
-We start by resolving global names, which are necessary for
-namespace names.  The global entities are `class A`, `class
-B`, `fun A`, `fun foo(...)`, and `fun foo()`; for each, we
-pick the shortest name that isn't claimed by anything else.
-We end up with the following selection:
-* `class A` -> `class_A`
-* `class B` -> `B`
-* `fun A` -> `fun_A`
-* `fun foo(...)` -> `fun_foo_takes_x_B`
-* `fun foo()` -> `fun_foo_takes_Unit`
-
-This lets us resolve all namespace names that depend on
-names in the program, namely for `class A` and `class B`.
-(Note that while `fun foo()` is a namespace, the name
-of that namespace is `local`.)
-
-We then resolve the local names per namespace.  The
-interesting case here is in `class B`, where the getter and
-setter of `x` collide.
-
-Every namespace-name pair is now unique, so we get the
-following code, with unique names.  (Note that we compile
-properties to fields or methods; we do so here for B.x to
-demonstrate.)
-```kotlin
-class class_A(val class_A_x: Int) {
-    fun class_A_foo(): Int = ...
-}
+1. **Pick sinks in the dependency graph.**  
+   Take all vertices with out-degree $0$ (no unresolved dependencies).
 
-class B {
-    fun B_prop_x_getter(): Int { ... }
-    fun B_prop_x_setter(param_v: Int) { ... }
-}
+2. **Fix the version indicated by the sink vertex.**  
+   For each sink $(x, v)$, we attempt to **assign version $v$** to $x$.  
+   The assignment is allowed iff **all** of the following hold:
+  - **No collisions** are created with already assigned strings.
+  - **Basename fairness:** no other name with the **same basename** has
+    already been assigned a **strictly shorter** version than $v$.
```

Comment:
Doesn't this contradicts your group assignment requirement? If you assign things by base names, no other name with the same basename can have an assignment yet.